### ISSUE TITLE
Broken socket accounting

### ISSUE THREAD

**Issue by keshonok** [natsys/tempesta#60]
When a socket is freed the kernel issues several warnings regarding socket accounting. These warning tell that from the kernel's pint of view some resources were not freed at the time the socket is destroyed.

It's not that Tempesta doesn't free those resources. However it appears that Tempesta is careless with resource accounting when moving an SKB from the receive queue to the send queue.

```
Loading Tempesta kernel modules...
[   10.038782] [tempesta] Initializing Tempesta FW kernel module...
[   10.039460] [tempesta] init: cfg_if
[   10.039847] [tempesta] init: http
[   10.040255] [tempesta] init: server
[   10.040655] [tempesta] init: client
[   10.041044] [tempesta] init: session
[   10.041439] [tempesta] init: connection
[   10.041860] [tempesta]   register cfg: cache
[   10.042343] [tempesta]   register cfg: sock_backend
[   10.042895] [tempesta]   register cfg: sock_frontend
[   10.057398] [tempesta] Registering new scheduler: dummy
Starting Tempesta...
[   10.059534] [tempesta] got state via sysctl: start
[   10.059930] [tempesta]   reading configuration file...
[   10.060348] [tempesta]   reading file: /home/natsys-lab/tfw-run/tempesta/etc/tempesta_fw.conf
[   10.061421] [tempesta]   file size: 2250 bytes
[   10.061775] [tempesta]   read by offset: 0
[   10.062137] [tempesta]   read by offset: 2250
[   10.062509] [tempesta] starting all modules...
[   10.062858] [tempesta]   parsing configuration and pushing it to modules...
[   10.063396] [tempesta]   spec handle: 'backend'
[   10.063748] [tempesta]   spec handle: 'listen'
[   10.064195] [tempesta]   created front-end socket: sk=ffff88003cc76040
[   10.064697] [tempesta]   use default entry: 'cache off;'
[   10.065099] [tempesta]   spec handle: 'cache'
[   10.065431] [tempesta]   use default entry: 'cache_size 262144;'
[   10.065880] [tempesta]   spec handle: 'cache_size'
[   10.066246] [tempesta]   use default entry: 'cache_dir /opt/tempesta/cache;'
[   10.066774] [tempesta]   spec handle: 'cache_dir'
[   10.067133] [tempesta]   starting modules...
[   10.067452] [tempesta]   mod_start(): cache
[   10.067767] [tempesta]   mod_start(): sock_backend
[   10.068139] [tempesta]   Starting thread: tfw_bconnd
[   10.068547] [tempesta]   mod_start(): sock_frontend
[   10.068919] [tempesta]   start listening on socket: sk=ffff88003cc76040
[   10.069429] [tempesta] modules are started
net.tempesta.state = start
done
done
Starting nginx: nginx.
[   14.069328] [tempesta] Connected to backend: 192.168.10.1:8080
[   16.943182] [sync_socket]   state_change: TCP_ESTABLISHED
[   16.944378] [tempesta]   New client socket ffff88003cd14980 (state=1)
[   16.945755] [tempesta]   Link new msg ffff88003d712040 with connection ffff88003d9c5500
[   16.946924] [tempesta]   Add skb ffff88003db7b5c0 to message ffff88003d712040
[   16.946924] [tempesta]   Received 62 client data bytes (GET / HTTP/1.1
[   16.946924] User-Agent: Wget/1.15 (linux-gnu)
[   16.946924] Accept: */*) on socket (conn=ffff88003d9c5500)
[   16.946924] [tempesta]   enter FSM at state 0
[   16.946924] [tempesta]   parser: Req_0(0:0): c=0x47(G), r=-2
[   16.946924] [tempesta]   parser: Req_Method(1:0): c=0x47(G), r=-2
[   16.946924] [tempesta]   parser: Req_MUSpace(2:0): c=0x2f(/), r=-2
[   16.946924] [tempesta]   parser: Req_UriAbsPath(6:0): c=0x20( ), r=-2
[   16.946924] [tempesta]   parser: Req_HttpVer(7:0): c=0x48(H), r=-2
[   16.946924] [tempesta]   parser: Req_EoL(15:0): c=0xd(.), r=-2
[   16.946924] [tempesta]   parser: Req_EoL(15:0): c=0xa(.), r=-2
[   16.946924] [tempesta]   parser: Req_Hdr(16:0): c=0x55(U), r=-2
[   16.946924] [tempesta]   parser: Req_HdrOther(92:0): c=0x55(U), r=-2
[   16.946924] [tempesta]   store header w/ ptr=ffff88003cce1892 len=33 flags=0 id=3
[   16.946924] [tempesta]   parser: Req_Hdr(16:0): c=0x41(A), r=-2
[   16.946924] [tempesta]   parser: Req_HdrOther(92:0): c=0x41(A), r=-2
[   16.946924] [tempesta]   store header w/ ptr=ffff88003cce18b5 len=11 flags=0 id=4
[   16.946924] [tempesta]   request parsed: len=62 parsed=62 msg_len=62 res=-1
[   16.946924] [tempesta]   postpone skb ffff88003db7b5c0
[   16.946924] [tempesta]   Add skb ffff88003db7b5c0 to message ffff88003d712040
[   16.946924] [tempesta]   Received 50 client data bytes (
[   16.946924] Host: 192.168.10.230
[   16.946924] Connection: Keep-Alive
[   16.946924] 
[   16.946924] ) on socket (conn=ffff88003d9c5500)
[   16.946924] [tempesta]   enter FSM at state 92
[   16.946924] [tempesta]   parser: Req_HdrOther(92:0): c=0xd(.), r=-2
[   16.946924] [tempesta]   store header w/ ptr=          (null) len=1 flags=0 id=4
[   16.946924] [tempesta]   parser: Req_Hdr(16:0): c=0x48(H), r=-2
[   16.946924] [tempesta]   parser: RGen_LWS(10000:21): c=0x20( ), r=-2
[   16.946924] [tempesta]   parser: RGen_LWS(10000:21): c=0x31(1), r=-2
[   16.946924] [tempesta]   enter FSM at state 21
[   16.946924] [tempesta]   parser: Req_HdrHostV(21:0): c=0x31(1), r=-2
[   16.946924] [tempesta]   enter FSM at state 1
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x31(1), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x39(9), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x32(2), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x2e(.), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x31(1), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x36(6), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x38(8), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x2e(.), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x31(1), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x30(0), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x2e(.), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x32(2), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x33(3), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0x30(0), r=-2
[   16.946924] [tempesta]   parser: Req_I_H(1:1): c=0xd(.), r=-2
[   16.946924] [tempesta]   parser: Req_I_H_EoL(3:1): c=0xd(.), r=-2
[   16.946924] [tempesta]   parser: Req_I_H_EoL(3:1): c=0xa(.), r=-2
[   16.946924] [tempesta]   parse header __req_parse_host: ret=16 plen=14 id=1
[   16.946924] [tempesta]   store header w/ ptr=ffff88003e33b08e len=20 flags=0 id=1
[   16.946924] [tempesta]   parser: Req_Hdr(16:0): c=0x43(C), r=-2
[   16.946924] [tempesta]   parser: Req_HdrC(22:0): c=0x6f(o), r=-2
[   16.946924] [tempesta]   parser: Req_HdrConnection(44:0): c=0x3a(:), r=-2
[   16.946924] [tempesta]   parser: RGen_LWS(10000:45): c=0x20( ), r=-2
[   16.946924] [tempesta]   parser: RGen_LWS(10000:45): c=0x4b(K), r=-2
[   16.946924] [tempesta]   enter FSM at state 45
[   16.946924] [tempesta]   parser: Req_HdrConnectionV(45:0): c=0x4b(K), r=-2
[   16.946924] [tempesta]   enter FSM at state 1
[   16.946924] [tempesta]   parser: I_Conn(1:1): c=0x4b(K), r=-2
[   16.946924] [tempesta]   parser: I_EoT(6:1): c=0xd(.), r=0
[   16.946924] [tempesta]   parser: I_EoL(7:1): c=0xd(.), r=0
[   16.946924] [tempesta]   parser: I_EoL(7:1): c=0xa(.), r=0
[   16.946924] [tempesta]   parser: Connection parsed: flags 0x2
[   16.946924] [tempesta]   parse header __parse_connection: ret=12 plen=10 id=0
[   16.946924] [tempesta]   store header w/ ptr=ffff88003e33b0a4 len=22 flags=0 id=0
[   16.946924] [tempesta]   parser: Req_Hdr(16:0): c=0xd(.), r=-2
[   16.946924] [tempesta]   parser: Req_HdrDone(93:0): c=0xa(.), r=-2
[   16.946924] [tempesta]   parse msg body: flags=0x2 content_length=0
[   16.946924] [tempesta]   request parsed: len=50 parsed=50 msg_len=50 res=0
[   16.946924] [tempesta]   GFSM return code 0
[   16.946924] 
[   16.946924] =============================================
[   16.946924] [ INFO: possible recursive locking detected ]
[   16.946924] 3.10.10-tempesta #8 Tainted: G           O
[   16.946924] ---------------------------------------------
[   16.946924] swapper/0/0 is trying to acquire lock:
[   16.946924]  (slock-AF_INET/1){+.-...}, at: [<ffffffffa00002a8>] ss_send+0x28/0x160 [sync_socket]
[   16.946924] 
[   16.946924] but task is already holding lock:
[   16.946924]  (slock-AF_INET/1){+.-...}, at: [<ffffffff813433bc>] tcp_v4_rcv+0x40c/0x8b0
[   16.946924] 
[   16.946924] other info that might help us debug this:
[   16.946924]  Possible unsafe locking scenario:
[   16.946924] 
[   16.946924]        CPU0
[   16.946924]        ----
[   16.946924]   lock(slock-AF_INET/1);
[   16.946924]   lock(slock-AF_INET/1);
[   16.946924] 
[   16.946924]  *** DEADLOCK ***
[   16.946924] 
[   16.946924]  May be due to missing lock nesting notation
[   16.946924] 
[   16.946924] 3 locks held by swapper/0/0:
[   16.946924]  #0:  (rcu_read_lock){.+.+..}, at: [<ffffffff812e6463>] __netif_receive_skb_core+0x63/0x590
[   16.946924]  #1:  (rcu_read_lock){.+.+..}, at: [<ffffffff8131d750>] ip_local_deliver_finish+0x40/0x150
[   16.946924]  #2:  (slock-AF_INET/1){+.-...}, at: [<ffffffff813433bc>] tcp_v4_rcv+0x40c/0x8b0
[   16.946924] 
[   16.946924] stack backtrace:
[   16.946924] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G           O 3.10.10-tempesta #8
[   16.946924]  ffffffff8155d440 ffff88003fc03828 ffffffff813bf35d ffff88003fc03898
[   16.946924]  ffffffff81073b96 ffffffff81072787 0000000000000007 00000000ffffffff
[   16.946924]  ffff880000000000 00000b005849e24f ffffffff00000000 ffffffff816f6640
[   16.946924] Call Trace:
[   16.946924]  <IRQ>  [<ffffffff813bf35d>] dump_stack+0x19/0x1b
[   16.946924]  [<ffffffff81073b96>] __lock_acquire+0xfa6/0x1240
[   16.946924]  [<ffffffff81072787>] ? trace_hardirqs_on_caller+0x1d7/0x210
[   16.946924]  [<ffffffff81074658>] lock_acquire+0x98/0xd0
[   16.946924]  [<ffffffffa00002a8>] ? ss_send+0x28/0x160 [sync_socket]
[   16.946924]  [<ffffffff813c30ca>] _raw_spin_lock_nested+0x3a/0x50
[   16.946924]  [<ffffffffa00002a8>] ? ss_send+0x28/0x160 [sync_socket]
[   16.946924]  [<ffffffffa00002a8>] ss_send+0x28/0x160 [sync_socket]
[   16.946924]  [<ffffffffa0049413>] ? tfw_sched_get_srv+0x49/0x4b [tempesta_fw]
[   16.946924]  [<ffffffffa0011c8d>] tfw_connection_send_srv+0xbf/0xc6 [tempesta_fw]
[   16.946924]  [<ffffffffa00153a2>] tfw_http_req_cache_cb+0x5d/0x5f [tempesta_fw]
[   16.946924]  [<ffffffffa000ddec>] tfw_cache_req_process+0x39/0x163 [tempesta_fw]
[   16.946924]  [<ffffffffa0015345>] ? tfw_http_adjust_resp+0x13/0x13 [tempesta_fw]
[   16.946924]  [<ffffffffa0015660>] tfw_http_req_process+0x2bc/0x32b [tempesta_fw]
[   16.946924]  [<ffffffffa0015a3b>] tfw_http_msg_process+0x43/0x5e [tempesta_fw]
[   16.946924]  [<ffffffffa0012d67>] tfw_gfsm_dispatch+0x41/0x43 [tempesta_fw]
[   16.946924]  [<ffffffffa0011e4e>] tfw_connection_recv+0x88/0x8a [tempesta_fw]
[   16.946924]  [<ffffffffa000005b>] ss_tcp_process_proto_skb+0x5b/0x90 [sync_socket]
[   16.946924]  [<ffffffffa0000167>] ss_tcp_process_skb+0xd7/0x170 [sync_socket]
[   16.946924]  [<ffffffffa0000eae>] ss_tcp_data_ready+0x10e/0x25e [sync_socket]
[   16.946924]  [<ffffffff8133425d>] tcp_data_queue+0x4dd/0xd10
[   16.946924]  [<ffffffff81337660>] tcp_rcv_established+0x560/0x620
[   16.946924]  [<ffffffff8134183c>] tcp_v4_do_rcv+0x1dc/0x4e0
[   16.946924]  [<ffffffff811eea42>] ? do_raw_spin_lock+0x82/0x140
[   16.946924]  [<ffffffff813433e7>] tcp_v4_rcv+0x437/0x8b0
[   16.946924]  [<ffffffff8131d7c5>] ip_local_deliver_finish+0xb5/0x150
[   16.946924]  [<ffffffff8131d750>] ? ip_local_deliver_finish+0x40/0x150
[   16.946924]  [<ffffffff8131df70>] ip_local_deliver+0x50/0x90
[   16.946924]  [<ffffffff8131dcde>] ip_rcv_finish+0x47e/0x580
[   16.946924]  [<ffffffff8131e26c>] ip_rcv+0x2bc/0x340
[   16.946924]  [<ffffffff812e6869>] __netif_receive_skb_core+0x469/0x590
[   16.946924]  [<ffffffff812e6463>] ? __netif_receive_skb_core+0x63/0x590
[   16.946924]  [<ffffffff812e69df>] __netif_receive_skb+0x4f/0x70
[   16.946924]  [<ffffffff812e7543>] netif_receive_skb+0xb3/0x140
[   16.946924]  [<ffffffff812d9be9>] ? __netdev_alloc_skb+0x49/0xb0
[   16.946924]  [<ffffffff81270fd4>] virtnet_poll+0x624/0x770
[   16.946924]  [<ffffffff812e8783>] net_rx_action+0xa3/0x1f0
[   16.946924]  [<ffffffff81036550>] __do_softirq+0xe0/0x1d0
[   16.946924]  [<ffffffff810367c9>] irq_exit+0x69/0xc0
[   16.946924]  [<ffffffff81003c5d>] do_IRQ+0x9d/0xc0
[   16.946924]  [<ffffffff813c3df2>] common_interrupt+0x72/0x72
[   16.946924]  <EOI>  [<ffffffff81023626>] ? native_safe_halt+0x6/0x10
[   16.946924]  [<ffffffff8100a8dd>] default_idle+0x1d/0x30
[   16.946924]  [<ffffffff8100b26a>] arch_cpu_idle+0xa/0x10
[   16.946924]  [<ffffffff810664ac>] cpu_startup_entry+0x15c/0x200
[   16.946924]  [<ffffffff813b79ad>] rest_init+0xcd/0xe0
[   16.946924]  [<ffffffff813b78e0>] ? csum_partial_copy_generic+0x170/0x170
[   16.946924]  [<ffffffff815d1da0>] start_kernel+0x34d/0x358
[   16.946924]  [<ffffffff815d185c>] ? repair_env_string+0x5e/0x5e
[   16.946924]  [<ffffffff815d1120>] ? early_idt_handlers+0x120/0x120
[   16.946924]  [<ffffffff815d155b>] x86_64_start_reservations+0x2a/0x2c
[   16.946924]  [<ffffffff815d164c>] x86_64_start_kernel+0xef/0xfc
[   16.946924] [sync_socket]   ss_send:89 entail skb=ffff88003db7b5c0 data_len=50 len=112
[   16.946924] [sync_socket]   ss_send:99 sk=ffff88003cd15280 is_queue_empty=0 tcp_send_head(sk)=ffff88003db7b5c0 sk->sk_state=1
[   17.025136] [tempesta]   Link new msg ffff88003d505040 with connection ffff88003d9c5548
[   17.025741] [tempesta]   Add skb ffff88003db7b9c0 to message ffff88003d505040
[   17.026278] [tempesta]   received 62 server data bytes (HTTP/1.1 200 OK
[   17.026278] Server: nginx/1.4.6 (Ubuntu)
[   17.026278] Date: Tue, 24 F) on socket (conn=ffff88003d9c5548)
[   17.027497] [tempesta]   enter FSM at state 0
[   17.027824] [tempesta]   parser: Resp_0(0:0): c=0x48(H), r=-2
[   17.028264] [tempesta]   parser: Resp_HttpVer(2:0): c=0x48(H), r=-2
[   17.028732] [tempesta]   parser: Resp_StatusCode(11:0): c=0x32(2), r=-2
[   17.029124] [tempesta]   parser: Resp_ReasonPhrase(12:1): c=0x20( ), r=-2
[   17.029124] [tempesta]   parser: Resp_Hdr(13:1): c=0x53(S), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrOther(87:1): c=0x65(e), r=-2
[   17.029124] [tempesta]   parser: Resp_Hdr(13:1): c=0x44(D), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrOther(87:1): c=0x61(a), r=-2
[   17.029124] [tempesta]   response parsed: len=62 parsed=62 res=-1
[   17.029124] [tempesta]   postpone skb ffff88003db7b9c0
[   17.029124] [tempesta]   Add skb ffff88003db7b9c0 to message ffff88003d505040
[   17.029124] [tempesta]   received 796 server data bytes (eb 2015 10:29:05 GMT
[   17.029124] Content-Type: text/html
[   17.029124] Content-Length: 612
[   17.029124] Last-Modified: Tue, 04 Mar 2014 11:46:45 GMT
[   17.029124] Connection: keep-alive
[   17.029124] ETag: "5315bd25-264"
[   17.029124] Accept-Ranges: bytes
[   17.029124] 
[   17.029124] <!DOCTYPE html>
[   17.029124] <html>
[   17.029124] <head>
[   17.029124] <title>Welcome to nginx!</title>
[   17.029124] <style>
[   17.029124]     body {
[   17.029124]         width: 35em;
[   17.029124]         margin: 0 auto;
[   17.029124]         font-family: Tahoma, Verdana, Arial, sans-serif;
[   17.029124]     }
[   17.029124] </style>
[   17.029124] </head>
[   17.029124] <body>
[   17.029124] <h1>Welcome to nginx!</h1>
[   17.029124] <p>If you see this page, the nginx web server is successfully installed and
[   17.029124] working. Further configuration is required.</p>
[   17.029124] 
[   17.029124] [tempesta]   enter FSM at state 87
[   17.029124] [tempesta]   parser: Resp_HdrOther(87:1): c=0x65(e), r=-2
[   17.029124] [tempesta]   parser: Resp_Hdr(13:1): c=0x43(C), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrC(14:1): c=0x6f(o), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrCo(28:1): c=0x6e(n), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrCon(29:1): c=0x74(t), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrCont(38:1): c=0x65(e), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrConte(39:1): c=0x6e(n), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrConten(40:1): c=0x74(t), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrContent(41:1): c=0x2d(-), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrContent_(42:1): c=0x54(T), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrOther(87:1): c=0x79(y), r=-2
[   17.029124] [tempesta]   parser: Resp_Hdr(13:1): c=0x43(C), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrC(14:1): c=0x6f(o), r=-2
[   17.029124] [tempesta]   parser: RGen_LWS(10000:49): c=0x20( ), r=-2
[   17.029124] [tempesta]   parser: RGen_LWS(10000:49): c=0x36(6), r=-2
[   17.029124] [tempesta]   enter FSM at state 49
[   17.029124] [tempesta]   parser: Resp_HdrContent_LengthV(49:0): c=0x36(6), r=-2
[   17.029124] [tempesta]   enter FSM at state 3
[   17.029124] [tempesta]   parser: I_ContLen(3:3): c=0x36(6), r=-2
[   17.029124] [tempesta]   parser: I_EoL(7:3): c=0xd(.), r=-2
[   17.029124] [tempesta]   parser: I_EoL(7:3): c=0xa(.), r=-2
[   17.029124] [tempesta]   parse header __parse_content_length: ret=5 plen=3 id=3
[   17.029124] [tempesta]   store header w/ ptr=ffff88003d505548 len=19 flags=1 id=3
[   17.029124] [tempesta]   parser: Resp_Hdr(13:0): c=0x4c(L), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrOther(87:0): c=0x61(a), r=-2
[   17.029124] [tempesta]   parser: Resp_Hdr(13:0): c=0x43(C), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrC(14:0): c=0x6f(o), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrConnection(36:0): c=0x3a(:), r=-2
[   17.029124] [tempesta]   parser: RGen_LWS(10000:37): c=0x20( ), r=-2
[   17.029124] [tempesta]   parser: RGen_LWS(10000:37): c=0x6b(k), r=-2
[   17.029124] [tempesta]   enter FSM at state 37
[   17.029124] [tempesta]   parser: Resp_HdrConnectionV(37:0): c=0x6b(k), r=-2
[   17.029124] [tempesta]   enter FSM at state 1
[   17.029124] [tempesta]   parser: I_Conn(1:1): c=0x6b(k), r=-2
[   17.029124] [tempesta]   parser: I_EoT(6:1): c=0xd(.), r=0
[   17.029124] [tempesta]   parser: I_EoL(7:1): c=0xd(.), r=0
[   17.029124] [tempesta]   parser: I_EoL(7:1): c=0xa(.), r=0
[   17.029124] [tempesta]   parser: Connection parsed: flags 0x2
[   17.029124] [tempesta]   parse header __parse_connection: ret=12 plen=10 id=0
[   17.029124] [tempesta]   store header w/ ptr=ffff88003d505588 len=22 flags=1 id=0
[   17.029124] [tempesta]   parser: Resp_Hdr(13:0): c=0x45(E), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrE(50:0): c=0x54(T), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrOther(87:0): c=0x61(a), r=-2
[   17.029124] [tempesta]   parser: Resp_Hdr(13:0): c=0x41(A), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrOther(87:0): c=0x63(c), r=-2
[   17.029124] [tempesta]   parser: Resp_Hdr(13:0): c=0xd(.), r=-2
[   17.029124] [tempesta]   parser: Resp_HdrDone(88:0): c=0xa(.), r=-2
[   17.029124] [tempesta]   parse msg body: flags=0x2 content_length=612
[   17.029124] [tempesta]   parser: Resp_Body(89:0): c=0x3c(<), r=-2
[   17.029124] [tempesta]   read body: to_read=612
[   17.029124] [tempesta]   parser: Resp_BodyReadChunk(92:0): c=0x3c(<), r=-2
[   17.029124] [tempesta]   response parsed: len=796 parsed=796 res=0
[   17.029124] [sync_socket]   ss_send:89 entail skb=ffff88003db7b9c0 data_len=796 len=858
[   17.029124] [sync_socket]   ss_send:99 sk=ffff88003cd14980 is_queue_empty=0 tcp_send_head(sk)=ffff88003db7b9c0 sk->sk_state=1
[   17.029124] [tempesta]   Free msg: ffff88003d712040
[   17.029124] [tempesta]   free skb ffff88003db7b5c0: truesize=4864 sk=ffff88003cd14980, destructor=ffffffff812d2ea0 users=2
[   17.029124] [tempesta]   Free msg: ffff88003d505040
[   17.029124] [tempesta]   free skb ffff88003db7b9c0: truesize=4864 sk=ffff88003cd15280, destructor=ffffffff812d2ea0 users=2
[   17.029124] [sync_socket]   state_change: TCP_CLOSE_WAIT
[   17.029124] [sync_socket]   received FIN, do active close
[   17.029124] [sync_socket]   Close socket ffff88003cd14980 (account=1)
[   17.029124] [tempesta]   Close socket ffff88003cd14980, conn=ffff88003d9c5500
[   17.029124] [tempesta]   Free msg:           (null)
[   17.029124] [tempesta]   Free connection: ffff88003d9c5500
[   17.029124] [tempesta]   Detach from peer: ffff88003d9c5548
[   17.029124] [tempesta]   Free session: ffff88003d518490
[   17.071291] ------------[ cut here ]------------
[   17.071649] WARNING: at net/core/stream.c:201 sk_stream_kill_queues+0x12b/0x160()
[   17.072198] Modules linked in: tfw_sched_dummy(O) tempesta_fw(O) tempesta_db(O) sync_socket(O)
[   17.072879] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G           O 3.10.10-tempesta #8
[   17.073450]  0000000000000009 ffff88003fc03ae0 ffffffff813bf35d ffff88003fc03b18
[   17.074033]  ffffffff8102ee8c ffff88003cd14980 ffff88003cd14be8 ffff88003cce2262
[   17.074613]  0000000000000000 ffff88003cce225a ffff88003fc03b28 ffffffff8102ef65
[   17.075165] Call Trace:
[   17.075280]  <IRQ>  [<ffffffff813bf35d>] dump_stack+0x19/0x1b
[   17.075280]  [<ffffffff8102ee8c>] warn_slowpath_common+0x5c/0x80
[   17.075280]  [<ffffffff8102ef65>] warn_slowpath_null+0x15/0x20
[   17.075280]  [<ffffffff812df3ab>] sk_stream_kill_queues+0x12b/0x160
[   17.075280]  [<ffffffff8132a793>] inet_csk_destroy_sock+0x83/0xe0
[   17.075280]  [<ffffffff8132ba3d>] tcp_done+0x7d/0x90
[   17.075280]  [<ffffffff813384b0>] tcp_rcv_state_process+0xa80/0xc50
[   17.075280]  [<ffffffff81341a4b>] tcp_v4_do_rcv+0x3eb/0x4e0
[   17.075280]  [<ffffffff811eea42>] ? do_raw_spin_lock+0x82/0x140
[   17.075280]  [<ffffffff813433e7>] tcp_v4_rcv+0x437/0x8b0
[   17.075280]  [<ffffffff8131d7c5>] ip_local_deliver_finish+0xb5/0x150
[   17.075280]  [<ffffffff8131d750>] ? ip_local_deliver_finish+0x40/0x150
[   17.075280]  [<ffffffff8131df70>] ip_local_deliver+0x50/0x90
[   17.075280]  [<ffffffff8131dcde>] ip_rcv_finish+0x47e/0x580
[   17.075280]  [<ffffffff8131e26c>] ip_rcv+0x2bc/0x340
[   17.075280]  [<ffffffff812e6869>] __netif_receive_skb_core+0x469/0x590
[   17.075280]  [<ffffffff812e6463>] ? __netif_receive_skb_core+0x63/0x590
[   17.075280]  [<ffffffff812e69df>] __netif_receive_skb+0x4f/0x70
[   17.075280]  [<ffffffff812e7543>] netif_receive_skb+0xb3/0x140
[   17.075280]  [<ffffffff812d9be9>] ? __netdev_alloc_skb+0x49/0xb0
[   17.075280]  [<ffffffff81270fd4>] virtnet_poll+0x624/0x770
[   17.075280]  [<ffffffff812e8783>] net_rx_action+0xa3/0x1f0
[   17.075280]  [<ffffffff81036550>] __do_softirq+0xe0/0x1d0
[   17.075280]  [<ffffffff810367c9>] irq_exit+0x69/0xc0
[   17.075280]  [<ffffffff81003c5d>] do_IRQ+0x9d/0xc0
[   17.075280]  [<ffffffff813c3df2>] common_interrupt+0x72/0x72
[   17.075280]  <EOI>  [<ffffffff8100a8c9>] ? default_idle+0x9/0x30
[   17.075280]  [<ffffffff81023626>] ? native_safe_halt+0x6/0x10
[   17.075280]  [<ffffffff8100a8dd>] default_idle+0x1d/0x30
[   17.075280]  [<ffffffff8100b26a>] arch_cpu_idle+0xa/0x10
[   17.075280]  [<ffffffff810664ac>] cpu_startup_entry+0x15c/0x200
[   17.075280]  [<ffffffff813b79ad>] rest_init+0xcd/0xe0
[   17.075280]  [<ffffffff813b78e0>] ? csum_partial_copy_generic+0x170/0x170
[   17.075280]  [<ffffffff815d1da0>] start_kernel+0x34d/0x358
[   17.075280]  [<ffffffff815d185c>] ? repair_env_string+0x5e/0x5e
[   17.075280]  [<ffffffff815d1120>] ? early_idt_handlers+0x120/0x120
[   17.075280]  [<ffffffff815d155b>] x86_64_start_reservations+0x2a/0x2c
[   17.075280]  [<ffffffff815d164c>] x86_64_start_kernel+0xef/0xfc
[   17.075280] ---[ end trace b5f21e053c44a402 ]---
```

**Comment by krizhanovsky** [comment]
The problem analyzing is not enough. The warning 

        WARNING: at net/core/stream.c:201 sk_stream_kill_queues+0x12b/0x160()

is produced by

       WARN_ON(sk->sk_forward_alloc)

So which exactly callback called by ```skb->destructor``` decrements the counter? Which callback is registered for ```skb->destructor``` when we process the skb in ss_tcp_process_data()?

Also why __kfree_skb() called anyway for the skb doesn't decrement the counter properly?

Please explain all the points.

**Comment by keshonok** [comment]
`sk->sk_forward_alloc` isn't decremented. It's incremented. Different callbacks may do that, such as `sk_stream_rfree()` or `sock_rfree()`. They may be assigned to `skb->destructor` by  `sk_stream_set_owner_r()` or `sk_set_owner_r()`. From the accounting point of view it doesn't really matter as they do the same kind of a job. What does matter is a general principle of what happens and how that happens. According to it, we need to adjust accounting when we remove an skb from the socket's receive queue as that skb had been accounted for at the time it had been put there. We get a warning (not an error) if we don't do that. The system takes care of that by setting up an approproate skb destructor. In the regular course of things that happens when an skb is deleted as the data that it held has been consumed by an application.

Similar thing happens on the egress path.

The `sk->sk_forward_alloc` counter is for socket's own accounting. It has nothing to do with skb accounting. An skb is freed by calling `__kfree_skb()` when it's no longer needed. An skb destructor is called when the skb is freed, and the destructor makes all final touches that are required.

**Comment by krizhanovsky** [comment]
The unclear thing is that __kfree_skb() calls skb->destructor() in skb_release_head_state() called by skb_release_all(). So why do we need to call skb->destructor() explicitly via skb_orphan() as in the patch?

**Comment by keshonok** [comment]
In the comment to the path it's said that in the regualar course of action socket accounting data is adjusted when an SKB is freed. As explained above, that happens as stream data is consumed by an application, and the data is no longer needed to hang around in the kernel.

That's not what happens in Tempesta though. In the Synchronous Sockets architecture we "consume" the stream data inside Tempesta by taking SKBs out of the receive queue of one socket and putting them into the send queue of another, totally different socket. `__kfree_skb()` is never called for the receive socket in this case, and so the destructor is not called as well. These SKBs are then "re-owned" by the egress socket while not being correctly "un-owned" from the ingress socket. That's where orphaning comes in, and that's why we need to call it. Otherwise we see a warning for broken accounting in the ingress socket.

**Comment by krizhanovsky** [comment]
Ok, thanks. It seems we reassign destructor callback when pass the skb to other socket, this is normally done by skb_set_owner_{r,w}(). I'll investigate the issue more as part of overall Synchronous Sockets architecture review.

**Comment by keshonok** [comment]
This issue is somewhat connected with #56. If some SKBs are not released or released incorrectly, then socket accounting may break.

**Comment by keshonok** [comment]
It turned out that the kernel was giving accounting warnings on a socket that didn't belong to Tempesta. Issue #96 was discovered while researching this accounting problem. This problem is not reproduced after #96 was fixed.

**Comment by keshonok** [comment]
Recent Tempesta execution traces show that this issue is still pending, and this time it is happening with Tempesta sockets, which is proven by extra specific printouts added to the Linux kernel code.

This issue is actually caused by irregular freeing of SKBs that make up HTTP messages processed by Tempesta. Socket accounting is managed by calls to `skb->destructor()` when SKBs are freed. When an SKB is not released as it should be, that keeps bytes occoupied by that SKB accounted for by the socket, which, in turn, causes these kernel warnings when the socket is released.

This issue will be solved as soon as all SKBs are released properly, which is the essense of issue #56.

**Comment by krizhanovsky** [comment]
Warning ```WARNING: at net/core/stream.c:201 sk_stream_kill_queues``` for current master isn't 100% reproducable due to earlier crashes on skb freeing. I see the warnings in 30-50% of runs.

**Comment by keshonok** [comment]
This issue no longer comes up as SKBs are freed correctly now (see #56).
